[pkwangjoo] iP

pkwangjoo opened the pull request:
## Treebot

Treebot frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is, 
1. download it from [here](https://github.com/pkwangjoo/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 👍 

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

****
If you are Java programer, you can use it to practice Java too. Here's the `main` method:

```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

pkwangjoo committed:
implement greet, echo and exit

pkwangjoo committed:
implement add, list

pkwangjoo committed:
implement mark as done

pkwangjoo committed:
implement ToDos, Events, Deadlines

pkwangjoo committed:
implement error handling

pkwangjoo committed:
implement delete task

pkwangjoo committed:
implement automated text ui testing

pkwangjoo committed:
remove redundant try catch block

pkwangjoo committed:
indent switch case according to convention

pkwangjoo committed:
implement save task list state to data file

pkwangjoo committed:
implement load tasks from saved data

pkwangjoo committed:
remove redundant print statement

pkwangjoo committed:
make treebot store date and time as LocalDateTime

pkwangjoo committed:
Merge branch 'branch-Level-7'

pkwangjoo committed:
Merge branch 'branch-Level-8'

pkwangjoo committed:
fix localDateTime formatting issue when casting task

pkwangjoo committed:
implement Ui class to abstract user interaction

pkwangjoo committed:
add commands for list, add, delete, mark, unmark

pkwangjoo committed:
abstract file r/w using storage class

pkwangjoo committed:
implement parser and integrate commands and storage

pkwangjoo committed:
implement tasklist to abstract task collection

pkwangjoo committed:
organise files into packages and delete unused imports

pkwangjoo committed:
Merge branch 'add-gradle-support'

pkwangjoo committed:
add junit tests

pkwangjoo committed:
delete original duke class

pkwangjoo committed:
implement find task by keyword

pkwangjoo committed:
add javadocs

pkwangjoo committed:
remove wildcard imports

pkwangjoo committed:
Merge branch 'branch-A-JavaDoc'

pkwangjoo committed:
Merge branch 'branch-A-CodingStandard'

pkwangjoo committed:
Merge branch 'branch-Level-9'

jefrai commented on src/main/java/treebot/TreeBot.java:
```diff
@@ -0,0 +1,49 @@
+
+                if (c.equals(EXIT_TOKEN)) {
+                    break;
+                }
```
The exit process is implemented incorrectly - the user input should be compared directly against `EXIT_TOKEN`.

```suggestion
                if (fullCommand.equals(EXIT_TOKEN)) {
                    break;
                }

                Command c = parser.parse(fullCommand);
```

jefrai commented on src/main/java/commands/ListCommand.java:
```diff
@@ -0,0 +1,13 @@
+
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage){
+        taskList.listTasks();
```
`ListCommand` prints no output when the `TaskList` is empty - consider printing line dividers / some specific output so that the user knows that their command has been received and processed successfully.

jefrai commented on src/main/java/tasks/TaskFactory.java:
```diff
@@ -0,0 +1,96 @@
+
+    private Task makeDeadline(String instruction) throws InvalidDateTimeFormatException {
+        String[] splitStr = instruction.split("\\s+", 2);
+        String taskDescription = splitStr[1].split(" /by ")[0];
```
Uncaught `ArrayIndexOutOfBoundsException` is thrown when no `/by` is supplied by the user, causing the program to crash - could catch it and print an error message to the user, and then continue to run normally.

Also, not sure if the `/by` token is being located correctly - `splitStr[1]` will only contain the second word of `instruction`. Perhaps just remove the first word of `instruction` and find `/by` from there?

jefrai commented on src/main/java/utils/Storage.java:
```diff
@@ -0,0 +1,92 @@
+     */
+    public Storage(String filePath)  {
+        this.f = new File(filePath);
+        this.filePath = filePath;
```
When the user first runs the program, the `/data` directory does not exist yet - we should create it in such a scenario to allow saving to work correctly.

```suggestion
        this.f = new File(filePath);
        f.getParentFile().mkdirs();
        this.filePath = filePath;
```

TimothyLawSongEn commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,19 @@
+import utils.Ui;
+
+public class FindCommand extends Command{
+    String keyword;
```
Should we add an extra line below for readability?

TimothyLawSongEn commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public String toStorageFormatString() {
+        return "D|" + (isDone ? "1" : "0") + "|" + this.taskDescription + "|" + this.deadline.format(DateTimeFormatter.ofPattern("d/M/yyyy HHmm"));
```
Should we provide a line break for this long statement?

TimothyLawSongEn commented on src/main/java/tasks/TaskFactory.java:
```diff
@@ -0,0 +1,96 @@
+package tasks;
+
```
Should we standardise the amount of line breaks between import statements?

TimothyLawSongEn commented on src/main/java/tasks/TaskFactory.java:
```diff
@@ -0,0 +1,96 @@
+
+        try {
+            switch (taskType) {
+            case "todo":
```
I like how you adhered to the Switch case coding standards.

TimothyLawSongEn commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,87 @@
+
+    /**
+     * Returns a copy of the ArrayList of Task.
+     * @return a copy of the ArrayList of Task.
```
Should a line break be added before this @return statement?

TimothyLawSongEn commented on src/main/java/utils/Parser.java:
```diff
@@ -0,0 +1,58 @@
+        switch (command) {
+        case "list":
+            return new ListCommand();
+        case "todo":
```
is this switch case required considering there is no implementation?

TimothyLawSongEn reviewed (commented):
LGTM. Just some nits to fix.

jefrai commented on src/main/java/utils/Parser.java:
```diff
@@ -0,0 +1,58 @@
+import commands.AddCommand;
+import commands.Command;
+import commands.DeleteCommand;
+import commands.ListCommand;
```
`Parser.java` uses `commands.FindCommand` and requires an import for it.

```suggestion
import commands.FindCommand;
import commands.ListCommand;
```

jefrai commented on src/main/java/commands/ToggleMarkTaskCommand.java:
```diff
@@ -0,0 +1,40 @@
+    @Override
+    public void execute(TaskList taskList, Ui u, Storage storage) throws TreeBotException {
+        if (!isMarkAsDone) {
+            taskList.unmarkTask(index);
```
`ArrayIndexOutOfBoundsException` could be thrown here - could also catch it and inform the user.

jefrai reviewed (commented):
No major issues in how the code is structured - good work!
However, there are currently several implementation details which could be refined:

pkwangjoo committed:
add missing imports

pkwangjoo committed:
make commands return result string

pkwangjoo committed:
create default constructor for treebot

pkwangjoo committed:
create main method in treebot class

pkwangjoo committed:
add basic gui for treebot

pkwangjoo committed:
Merge branch 'branch-Level-10'

pkwangjoo committed:
Command Classes: Add assertions

execute() function within each instances of commands perform
certain actions and contains toResultString() subroutine to
convert actions done to printable string for the UI to show
the user what was done.

There is strong coupling between the two functions
as the context of the command is being stored as attributes within
the command classes by execute()) (e.g. taskList, deletedTask)) and is
used within the toResultString() subroutine.

Creating assertions within toResultString() subroutine
will ensure that the result of the execution of command is returned as a
string and is correct by ensuring the assumption that the context is
stored by the command classes before execution of relevant action, after
which the result can be returned in a string format.

pkwangjoo committed:
Merge pull request #2 from pkwangjoo/branch-A-Assertions

Command Classes: Add assertions

pkwangjoo committed:
Delete all text based ui code

pkwangjoo committed:
Implement undo feature
Commands: Inject context before execution

pkwangjoo committed:
Merge branch 'branch-C-Undo'

pkwangjoo committed:
Replace the user and treeBot image and resize dialog box

pkwangjoo committed:
Merge branch 'branch-A-BetterGui'

pkwangjoo committed:
TaskFactory: Add default throw statement

pkwangjoo committed:
TaskFactory: Refactor to prevent parsing taskType again

pkwangjoo committed:
Add line breaks for better readability

pkwangjoo committed:
Add more javadocs

pkwangjoo committed:
Add more javadocs and error handling

pkwangjoo committed:
Parser: Add more error handling

pkwangjoo committed:
Greet user in gui

pkwangjoo committed:
Refactor packages and implement exit command

pkwangjoo committed:
Refactor the code to SLAP

pkwangjoo committed:
Storage: Add more unit test cases

pkwangjoo committed:
EventProps: Remove type parameter

pkwangjoo committed:
Add more tests and error handling

pkwangjoo committed:
Merge branch 'branch-A-CodeQuality'

pkwangjoo committed:
Refine testcases and add index error handling for mark task

pkwangjoo committed:
Storage: Make directory and file if not found
Update javadocs

pkwangjoo committed:
Add ui.png

pkwangjoo committed:
Add user guide

pkwangjoo committed:
Parser: Fix add todo parsing truncating task description

pkwangjoo committed:
Set main class name in build.gradle

pkwangjoo committed:
Add release download link to user guide
